### PR TITLE
Cleanup legacy JIT

### PR DESCRIPTION
**Pull request by bdw**
Branch `even-moar-jit` contained numerous improvements and cleanups that affected the legacy JIT. Having these separate would mean that even-moar-jit could not not merge cleanly without developer involvement because changes to the legacy JIT would mean it diverted.
This branch separates out the legacy JIT changes from even-moar-jit, so that even-moar-jit can be rebased cleanly on top of them, and we have a simpler patch set to review.

### CODE REVIEW

File: src/jit/compile.c [hunk 1]
```diff
@@ -1,83 +1,120 @@
...
 #include "platform/mmap.h"
-#include "emit.h"
 
-#define COPY_ARRAY(a, n, t) memcpy(MVM_malloc(n * sizeof(t)), a, n * sizeof(t))
+
+void MVM_jit_compiler_init(MVMThreadContext *tc, MVMJitCompiler *compiler, MVMJitGraph *jg);
+void MVM_jit_compiler_deinit(MVMThreadContext *tc, MVMJitCompiler *compiler);
+MVMJitCode * MVM_jit_compiler_assemble(MVMThreadContext *tc, MVMJitCompiler *compiler, MVMJitGraph *jg);
+
+#define COPY_ARRAY(a, n) memcpy(MVM_malloc(n * sizeof(a[0])), a, n * sizeof(a[0]))
```
Comment (jnthn) [inline src/jit/compile.c]: This would be preferable as a `MVM_STATIC_INLINE` rather than a macro.

File: src/jit/compile.c [hunk 2]
```diff
@@ -88,61 +125,48 @@ MVMJitCode * MVM_jit_compile_graph(MVMThreadContext *tc, MVMJitGraph *jg) {
...
-    code->inlines      = code->num_inlines ? COPY_ARRAY(jg->inlines, jg->num_inlines, MVMJitInline) : NULL;
-
-    /* clear up the assembler */
-    dasm_free(&state);
-    MVM_free(dasm_globals);
+    /* Copy the deopts, inlines, and handlers. Because these use the
+     * label index rather than the direct pointer, no fixup is
+     * necessary */
+    code->num_deopts   = jg->deopts_num;
+    code->deopts       = code->num_deopts ? COPY_ARRAY(jg->deopts, jg->deopts_num) : NULL;
```
Comment (jnthn) [inline src/jit/compile.c]: Wonder if we should fold the zero check into `COPY_ARRAY` rather than repeating it here?

File: src/jit/compile.c [hunk 3]
```diff
@@ -88,61 +125,48 @@ MVMJitCode * MVM_jit_compile_graph(MVMThreadContext *tc, MVMJitGraph *jg) {
...
+    code->num_handlers = jg->handlers_num;
+    code->handlers     = code->num_handlers ? COPY_ARRAY(jg->handlers, jg->handlers_alloc) : NULL;
+    code->num_inlines  = jg->inlines_num;
+    code->inlines      = code->num_inlines ? COPY_ARRAY(jg->inlines, jg->inlines_alloc) : NULL;
 
-    code->seq_nr = MVM_incr(&tc->instance->jit_seq_nr);
+    /* add sequence number */
+    code->seq_nr       = MVM_incr(&tc->instance->jit_seq_nr);
```
Comment (jnthn) [inline src/jit/compile.c]: Since the JIT now runs on the spesh thread only, an atomic increment ain't needed here.

File: src/jit/compile.h [hunk 4]
```diff
@@ -1,20 +1,18 @@
...
+typedef void (*MVMJitFunc)(MVMThreadContext *tc, MVMCompUnit *cu, void * label);
 
 struct MVMJitCode {
     MVMJitFunc func_ptr;
     size_t     size;
     MVMuint8  *bytecode;
 
     MVMStaticFrame *sf;
-    /* The basic idea here is that /all/ label names are indexes into
+    /* The basic igdea here is that /all/ label names are indexes into
```
Comment (jnthn) [inline src/jit/compile.h]: I'm pretty sure it was an idea...

File: src/jit/pseudo.c [hunk 5]
```diff
@@ -0,0 +1,490 @@
+#include "moar.h"
```
Comment (jnthn) [inline src/jit/pseudo.c]: Was this file meant to be included in this PR? It looks expr-JIT related.

**Comment by jnthn** [comment]
Looks very good overall; left a few comments.

File: src/jit/pseudo.c (hunk 5 above)
Comment (bdw) [reply src/jit/pseudo.c]: Nope, not sure how it got in, will remove it :-)

File: src/jit/compile.c (hunk 3 above)
Comment (bdw) [reply src/jit/compile.c]: Agreed, and i was actually thinking of moving the assignment of the sequence number to spesh graph time. Will do that in some other commit though.

File: src/jit/compile.c (hunk 1 above)
Comment (bdw) [reply src/jit/compile.c]: In that case it couldn't 'automatically' take over the types, which the macro can; it's really just a convenience though, so I don't care that much.

Comment (jnthn) [reply src/jit/compile.c]: Ah, good point. Can stay as a macro then.

**Pull request merged**